### PR TITLE
StorageConsumer: [CephFS] Add controller changes to provide storage service for consumers 

### PR DESCRIPTION
**Pull request by kesavanvt**

### CODE REVIEW

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 1]
```diff
@@ -52,14 +80,582 @@ func (r *StorageConsumerReconciler) Reconcile(ctx context.Context, request recon
...
+
+func (r *StorageConsumerReconciler) reconcileCephClientCephFSProvisioner() error {
+
+	// TODO: Update caps relavent to cephfs provisioner
+	desired := &rookCephv1.CephClient{
+		Spec: rookCephv1.ClientSpec{
+			Caps: map[string]string{
+				"mon": "allow r",
+				"mgr": "allow rw",
+				"osd": fmt.Sprintf("allow rw tag cephfs metadata=* path=%s", r.cephFilesystemSubVolumeGroup.Name),
```
Comment (kesavanvt) [inline controllers/storageconsumer/storageconsumer_controller.go:464]: Rook uses 
```
"mon", "allow r", 
"mgr", "allow rw",
"osd", "allow rw tag cephfs metadata=*"
```

@leseb Can you check the cap changes done is correct for subvolume group integration

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 2]
```diff
@@ -52,14 +80,582 @@ func (r *StorageConsumerReconciler) Reconcile(ctx context.Context, request recon
...
+func (r *StorageConsumerReconciler) reconcileCephClientCephFSNode() error {
+
+	// TODO: Update caps relavent to cephfs node
+	desired := &rookCephv1.CephClient{
+		Spec: rookCephv1.ClientSpec{
+			Caps: map[string]string{
+				"mon": "allow r",
+				"mgr": "allow rw",
+				"osd": fmt.Sprintf("allow rw tag cephfs *=* path=%s", r.cephFilesystemSubVolumeGroup.Name),
+				"mds": fmt.Sprintf("allow rw path=%s", r.cephFilesystemSubVolumeGroup.Name),
```
Comment (kesavanvt) [inline controllers/storageconsumer/storageconsumer_controller.go:505]: Rook Uses 
```
"mon", "allow r",
"mgr", "allow rw",
"osd", "allow rw tag cephfs *=*",
"mds", "allow rw"
```
@leseb Can you check the cap changes done is correct for subvolume group integration

Comment (leseb) [reply controllers/storageconsumer/storageconsumer_controller.go:505]: @ajarr please confirm that the subvolume group integration (`path`) is correctly written. Thanks!

File: controllers/storageconsumer/storageconsumer_controller.go (hunk 1 above)
Comment (leseb) [reply controllers/storageconsumer/storageconsumer_controller.go:464]: @ajarr please confirm that the subvolume group integration (`path`) is correctly written. Thanks!

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 3]
```diff
@@ -284,6 +319,63 @@ func (r *StorageConsumerReconciler) reconcileCephBlockPool() error {
...
 }
 
+func (r *StorageConsumerReconciler) reconcileCephFilesystemSubVolumeGroup() error {
+
+	cephFilesystemList := rookCephv1.CephFilesystemList{}
+	err := r.Client.List(r.ctx, &cephFilesystemList, client.InNamespace(r.namespace))
+
+	if err != nil {
+		return fmt.Errorf("Error fetching CephFilesystemList. %+v", err)
+	}
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:329]:
```suggestion
	if err := r.Client.List(r.ctx, &cephFilesystemList, client.InNamespace(r.namespace)); err != nil {
		return fmt.Errorf("Error fetching CephFilesystemList. %+v", err)
	}
```

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 4]
```diff
@@ -284,6 +319,63 @@ func (r *StorageConsumerReconciler) reconcileCephBlockPool() error {
...
+
+	var cephFileSystemName string
+	availableCephFileSystems := len(cephFilesystemList.Items)
+	if availableCephFileSystems == 0 {
+		return fmt.Errorf("No CephFileSystem found in the cluster")
+	} else if availableCephFileSystems > 1 {
+		klog.Warningf("More than 1 CephFileSystem found in the cluster")
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:336]:
```suggestion
		klog.Warningf("More than one CephFileSystem found in the cluster, selecting the first one")
```

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 5]
```diff
@@ -164,6 +184,14 @@ func (r *StorageConsumerReconciler) reconcilePhases() (reconcile.Result, error)
...
+			return reconcile.Result{}, err
+		}
+
+		if err := r.reconcileCephClientCephFSNode(); err != nil {
+			return reconcile.Result{}, err
+		}
+
 		if err := r.reconcileCephClientHealthChecker(); err != nil {
 			return reconcile.Result{}, err
 		}
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:197]: Can we rearrange the reconciles based on concerns,  RBD then CephFS (or the other way around)? 
One way will be:
```
r.reconcileCephClientRBDProvisioner
r.reconcileCephClientRBDNode
r.reconcileCephBlockPool
```

Then 
```
r.reconcileCephClientCephFSProvisioner
r.reconcileCephClientCephFSNode
r.reconcileCephFilesystemSubVolumeGroup
```

Then 
```
r.reconcileCephClientHealthChecker
```

This way we can guarantee that even if we have an error in CephFS we still have functioning Block storage and on top of that the code will be more ready for the next step where we will provide functionality based on request/caps

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 6]
```diff
@@ -132,6 +140,14 @@ func (r *StorageConsumerReconciler) initReconciler(request reconcile.Request) {
 	r.cephClientRBDNode.Name = fmt.Sprintf("%s-%s", "cephclient-rbd-node", r.storageConsumer.Name)
 	r.cephClientRBDNode.Namespace = r.namespace
 
+	r.cephClientCephFSProvisioner = &rookCephv1.CephClient{}
+	r.cephClientCephFSProvisioner.Name = fmt.Sprintf("%s-%s", "cephuser-cephfs-provisioner", r.storageConsumer.Name)
+	r.cephClientCephFSProvisioner.Namespace = r.namespace
+
+	r.cephClientCephFSNode = &rookCephv1.CephClient{}
+	r.cephClientCephFSNode.Name = fmt.Sprintf("%s-%s", "cephuser-cephfs-node", r.storageConsumer.Name)
+	r.cephClientCephFSNode.Namespace = r.namespace
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:149]: Is there a reason why the new clients are prefixed with `cephuser` while the old ones are prefixed with `cephclient`?

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 7]
```diff
@@ -284,6 +319,61 @@ func (r *StorageConsumerReconciler) reconcileCephBlockPool() error {
...
+	}
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.cephFilesystemSubVolumeGroup, func() error {
+		if err := r.own(r.cephFilesystemSubVolumeGroup); err != nil {
+			return err
+		}
+
+		r.cephFilesystemSubVolumeGroup.Spec = desired.Spec
+		return nil
+	})
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:352]: Why create `rookCephv1.CephFilesystemSubVolumeGroup` object if we are only using the spec and also why outside of the CreateOrUpdate call? 


```suggestion
	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.cephFilesystemSubVolumeGroup, func() error {
		if err := r.own(r.cephFilesystemSubVolumeGroup); err != nil {
			return err
		}

		r.cephFilesystemSubVolumeGroup.Spec = rookCephv1.CephFilesystemSubVolumeGroupSpec{
			FilesystemName: cephFileSystemName,
		}
		return nil
	})
```

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 8]
```diff
@@ -284,6 +319,61 @@ func (r *StorageConsumerReconciler) reconcileCephBlockPool() error {
...
+		},
+	}
+	if r.cephFilesystemSubVolumeGroup.Status != nil {
+		cephFilesystemSubVolumeGroupDetails.Phase = string(r.cephFilesystemSubVolumeGroup.Status.Phase)
+	}
+
+	r.updateCephResourceStatus(cephFilesystemSubVolumeGroupDetails)
+
+	if err != nil {
+		r.Log.Error(err, "Failed to update CephFilesystemSubVolumeGroup.", "CephFilesystemSubVolumeGroup", klog.KRef(r.cephFilesystemSubVolumeGroup.Namespace, r.cephFilesystemSubVolumeGroup.Name))
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:369]: This line is too long, please split over multiple lines:
```suggestion
		r.Log.Error(
			err, 
			"Failed to update CephFilesystemSubVolumeGroup.", 
			"CephFilesystemSubVolumeGroup", 
			klog.KRef(r.cephFilesystemSubVolumeGroup.Namespace, r.cephFilesystemSubVolumeGroup.Name),
		)
```

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 9]
```diff
@@ -361,6 +451,87 @@ func (r *StorageConsumerReconciler) reconcileCephClientRBDNode() error {
...
+	}
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.cephClientCephFSProvisioner, func() error {
+		if err := r.own(r.cephClientCephFSProvisioner); err != nil {
+			return err
+		}
+
+		r.cephClientCephFSProvisioner.Spec = desired.Spec
+		return nil
+	})
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:474]: Same comment around desired state and organization of the code

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 10]
```diff
@@ -361,6 +451,87 @@ func (r *StorageConsumerReconciler) reconcileCephClientRBDNode() error {
...
+		Name: r.cephClientCephFSProvisioner.Name,
+	}
+	if r.cephClientCephFSProvisioner.Status != nil {
+		cephClientCephFSProvisionerDetails.Phase = string(r.cephClientCephFSProvisioner.Status.Phase)
+	}
+
+	r.updateCephResourceStatus(cephClientCephFSProvisionerDetails)
+
+	if err != nil {
+		r.Log.Error(err, "Failed to update CephClient.", "CephClient", klog.KRef(r.cephClientCephFSProvisioner.Namespace, r.cephClientCephFSProvisioner.Name))
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:487]: Line too long, see prev comment on the same subject

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 11]
```diff
@@ -361,6 +451,87 @@ func (r *StorageConsumerReconciler) reconcileCephClientRBDNode() error {
...
+	}
+
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.cephClientCephFSNode, func() error {
+		if err := r.own(r.cephClientCephFSNode); err != nil {
+			return err
+		}
+
+		r.cephClientCephFSNode.Spec = desired.Spec
+		return nil
+	})
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:515]: Same comment around desired state and organization of the code

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 12]
```diff
@@ -361,6 +451,87 @@ func (r *StorageConsumerReconciler) reconcileCephClientRBDNode() error {
...
+		Name: r.cephClientCephFSNode.Name,
+	}
+	if r.cephClientCephFSNode.Status != nil {
+		cephClientCephFSNodeDetails.Phase = string(r.cephClientCephFSNode.Status.Phase)
+	}
+
+	r.updateCephResourceStatus(cephClientCephFSNodeDetails)
+
+	if err != nil {
+		r.Log.Error(err, "Failed to update CephClient.", "CephClient", klog.KRef(r.cephClientCephFSNode.Namespace, r.cephClientCephFSNode.Name))
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:528]: Line too long, see prev comment on the same subject

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 13]
```diff
@@ -284,6 +319,63 @@ func (r *StorageConsumerReconciler) reconcileCephBlockPool() error {
...
+		return fmt.Errorf("Error fetching CephFilesystemList. %+v", err)
+	}
+
+	var cephFileSystemName string
+	availableCephFileSystems := len(cephFilesystemList.Items)
+	if availableCephFileSystems == 0 {
+		return fmt.Errorf("No CephFileSystem found in the cluster")
+	} else if availableCephFileSystems > 1 {
+		klog.Warningf("More than one CephFileSystem found in the cluster, selecting the first one")
+	}
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:335]: No need for `else` as the prev if is returning

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 14]
```diff
@@ -284,6 +319,63 @@ func (r *StorageConsumerReconciler) reconcileCephBlockPool() error {
...
+
+	r.updateCephResourceStatus(cephFilesystemSubVolumeGroupDetails)
+
+	if err != nil {
+		r.Log.Error(
+			err,
+			"Failed to update CephFilesystemSubVolumeGroup.",
+			"CephFilesystemSubVolumeGroup",
+			klog.KRef(r.cephFilesystemSubVolumeGroup.Namespace,
+				r.cephFilesystemSubVolumeGroup.Name),
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:370]: As of this code base conventions, we do not break lines in this manner:
Please change to one of the following:
```
			klog.KRef(r.cephFilesystemSubVolumeGroup.Namespace, r.cephFilesystemSubVolumeGroup.Name),
```
or 
```
			klog.KRef(
				r.cephFilesystemSubVolumeGroup.Namespace,
				r.cephFilesystemSubVolumeGroup.Name,
			),
```

The main reason is that git's diff tracking is line-based. as such we either have the entire param on a single line of split the call to line per param

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 15]
```diff
@@ -361,6 +453,83 @@ func (r *StorageConsumerReconciler) reconcileCephClientRBDNode() error {
...
+
+	r.updateCephResourceStatus(cephClientCephFSNodeDetails)
+
+	if err != nil {
+		r.Log.Error(
+			err,
+			"Failed to update CephClient.",
+			"CephClient",
+			klog.KRef(r.cephClientCephFSNode.Namespace,
+				r.cephClientCephFSNode.Name),
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:525]: See comment above about breaking lines

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 16]
```diff
@@ -284,6 +319,63 @@ func (r *StorageConsumerReconciler) reconcileCephBlockPool() error {
...
+			"node":        r.cephClientCephFSNode.Name,
+		},
+	}
+	if r.cephFilesystemSubVolumeGroup.Status != nil {
+		cephFilesystemSubVolumeGroupDetails.Phase = string(r.cephFilesystemSubVolumeGroup.Status.Phase)
+	}
+
+	r.updateCephResourceStatus(cephFilesystemSubVolumeGroupDetails)
+
+	if err != nil {
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:364]: Why are we dealing with the error after the status update? If there was an error it usually mean that the resource was not created or updated, in that case, we should not update the status.

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 17]
```diff
@@ -361,6 +453,83 @@ func (r *StorageConsumerReconciler) reconcileCephClientRBDNode() error {
...
+		Kind: r.cephClientCephFSProvisioner.Kind,
+		Name: r.cephClientCephFSProvisioner.Name,
+	}
+	if r.cephClientCephFSProvisioner.Status != nil {
+		cephClientCephFSProvisionerDetails.Phase = string(r.cephClientCephFSProvisioner.Status.Phase)
+	}
+
+	r.updateCephResourceStatus(cephClientCephFSProvisionerDetails)
+
+	if err != nil {
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:483]: See comment above about error handling for `createOrUpdate`

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 18]
```diff
@@ -361,6 +453,83 @@ func (r *StorageConsumerReconciler) reconcileCephClientRBDNode() error {
...
+		Kind: r.cephClientCephFSNode.Kind,
+		Name: r.cephClientCephFSNode.Name,
+	}
+	if r.cephClientCephFSNode.Status != nil {
+		cephClientCephFSNodeDetails.Phase = string(r.cephClientCephFSNode.Status.Phase)
+	}
+
+	r.updateCephResourceStatus(cephClientCephFSNodeDetails)
+
+	if err != nil {
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:519]: See comment above about error handling for `createOrUpdate`

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 19]
```diff
@@ -392,7 +561,13 @@ func (r *StorageConsumerReconciler) reconcileCephClientHealthChecker() error {
 	r.updateCephResourceStatus(cephClientHealthCheckerDetails)
 
 	if err != nil {
-		r.Log.Error(err, "Failed to update CephClient.", "CephClient", klog.KRef(r.cephClientHealthChecker.Namespace, r.cephClientHealthChecker.Name))
+		r.Log.Error(
+			err,
+			"Failed to update CephClient.",
+			"CephClient",
+			klog.KRef(r.cephClientHealthChecker.Namespace,
+				r.cephClientHealthChecker.Name),
```
Comment (nb-ohad) [inline controllers/storageconsumer/storageconsumer_controller.go:569]: See comment above about breaking lines

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/storageconsumer/storageconsumer_controller.go (hunk 19 above)
Comment (kesavanvt) [reply controllers/storageconsumer/storageconsumer_controller.go]: Health checker changes will be done in a seperate PR, just keeping only the cephfs changes

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 20]
```diff
@@ -392,7 +561,13 @@ func (r *StorageConsumerReconciler) reconcileCephClientHealthChecker() error {
 	r.updateCephResourceStatus(cephClientHealthCheckerDetails)
 
 	if err != nil {
```
[the remark this answers is not in the document]
Comment (kesavanvt) [reply controllers/storageconsumer/storageconsumer_controller.go]: Health checker changes will be done in a seperate PR, just keeping only the cephfs changes

**Review by subhamkrai — APPROVED** [review]

File: controllers/storageconsumer/storageconsumer_controller.go (hunk 1 above)
Comment (vshankar) [reply controllers/storageconsumer/storageconsumer_controller.go:467]: The `path=<>` is a part of MDS cap specification, not the OSD. See here:

        https://docs.ceph.com/en/latest/cephfs/client-auth/

e.g.:

        caps: [mds] allow r, allow rw path=/bar

where `path` in your case would be the path to the subvolume or subvolume group as required.

File: controllers/storageconsumer/storageconsumer_controller.go [hunk 21]
```diff
@@ -361,6 +453,87 @@ func (r *StorageConsumerReconciler) reconcileCephClientRBDNode() error {
...
+	_, err := ctrl.CreateOrUpdate(r.ctx, r.Client, r.cephClientCephFSNode, func() error {
+		if err := r.own(r.cephClientCephFSNode); err != nil {
+			return err
+		}
+
+		r.cephClientCephFSNode.Spec = rookCephv1.ClientSpec{
+			Caps: map[string]string{
+				"mon": "allow r",
+				"mgr": "allow rw",
+				"osd": fmt.Sprintf("allow rw tag cephfs *=* path=%s", r.cephFilesystemSubVolumeGroup.Name),
```
Comment (vshankar) [inline controllers/storageconsumer/storageconsumer_controller.go:507]: Providing `path=<>` to the OSD does not make sense - file system "path" is a part of MDS (cap spec) as done below for mds caps.

**Pull request merged**